feat(subscriptions): Implement `SubscriptionResultCodec`

wedamija opened the pull request:
Just to get a feel for if this makes sense, will tidy things up a bit if so.

wedamija committed:
feat(subscriptions): Rough draft of `SubscriptionResultProducer` and `SubscriptionResultCodec`

Just to get a feel for if this makes sense, will tidy things up a bit if so.

tkaemming commented on snuba/utils/streams/kafka.py:
```diff
@@ -697,7 +697,7 @@ class KafkaProducer(Producer[TPayload]):
     def __init__(
         self, configuration: Mapping[str, Any], codec: Codec[KafkaPayload, TPayload]
     ) -> None:
-        self.__configuretion = configuration
```
omg

tkaemming reviewed (commented):
Lots of thoughts inline, some more useful than others, but the general ideas:

1. We don't need a concrete producer class, the codec itself should be fine.
2. We should spend a bit more time thinking about how we deal with errors when running these queries (does a bad apple spoil the whole bunch, how to deal with timeouts, etc.)
3. We probably don't want to serialize the `Future` directly since we can't deserialize them later (and trying to would probably expose implementation details that are not useful to consumers.)
4. We should revise the output data structure before we get locked into it.

wedamija commented on snuba/subscriptions/producer.py:
```diff
@@ -0,0 +1,6 @@
+
+
+class SubscriptionResultProducer(KafkaProducer[SubscriptionResult]):
+    pass
```
Yeah that makes sense, just still getting used to using generics vs subclassing all the time

wedamija commented on snuba/subscriptions/scheduler.py:
```diff
@@ -1,31 +1,14 @@
-    timestamp: datetime
-
-    # The task that should be executed.
-    task: TTask
```
I had some circular import issues here. This might be less of an issue after other changes, we'll see.

wedamija commented on snuba/subscriptions/codecs.py:
```diff
@@ -26,3 +32,48 @@ def decode(self, value: bytes) -> SubscriptionData:
+            # TODO: Not sure if this is right, but we likely want to rely on the same
+            # partition logic as elsewhere, and I believe we partition on project_id as
+            # key
+            str(value.task.task.data.project_id).encode("utf-8"),
```
No strong opinions about how we partition these, the consumer on the sentry side doesn't particularly care. I think subscription identifier is probably best.

wedamija commented on snuba/subscriptions/codecs.py:
```diff
@@ -26,3 +32,48 @@ def decode(self, value: bytes) -> SubscriptionData:
+    def decode(self, value: KafkaPayload) -> SubscriptionResult:
+        # Not really possible to decode this back into a `SubscriptionResult`, should
+        # we have a different type of codec that expects this?
+        return json.loads(value.value.decode("utf-8"))
```
1. Just the subscription id is fine?
3. Are you thinking just the conditions from the subscription? Should we just include the entire serialized subscription?

I think it makes sense to make this new type. I guess that whatever calls `.execute` will be responsible for converting the `ClickhouseQueryResult` into this format, and to handle exceptions etc.

wedamija committed:
Remove Producer subclass
Add `QueryResult` dataclass and modified codec to work with it
Removed `SubscriptionResult` since it's not relevant to this pr

wedamija committed:
tidying

wedamija committed:
more tidying

tkaemming reviewed (changes requested):
A few minor requests:

1. Please rebase (or merge) this (so that you can use `SubscriptionResult` from #743.)
2. Either provide a way to transform `SubscriptionResult` into `QueryResult` or use `SubscriptionResult` directly in the codec. It doesn't need to be bidirectional right now unless you really want to make it so. We're going to need to change it soon anyway to handle error results.
3. Along those lines, let's remove the tick derived data (partition and offset, maybe interval — I'm not sure what that represents) for now.
4. Let's return the `ClickhouseQueryResult` as the result/values type so this has a consistent structure with the HTTP interface.

wedamija committed:
Merge branch 'master' into danf/subscription_result_producer

wedamija committed:
Update codec

wedamija committed:
Merge branch 'master' into danf/subscription_result_producer

tkaemming commented on snuba/subscriptions/worker.py:
```diff
@@ -47,3 +50,24 @@ def flush_batch(self, batch: Sequence[Sequence[SubscriptionResult]]) -> None:
+    def encode(self, value: SubscriptionResult) -> KafkaPayload:
+        subscription_id = str(value.task.task.identifier)
+        return KafkaPayload(
+            str(subscription_id).encode("utf-8"),
```
```suggestion
            subscription_id.encode("utf-8"),
```

This is already a str due to the coerce on line 57.

tkaemming commented on snuba/subscriptions/worker.py:
```diff
@@ -47,3 +50,24 @@ def flush_batch(self, batch: Sequence[Sequence[SubscriptionResult]]) -> None:
+                {
+                    "version": 1,
+                    "payload": {
+                        "subscription_id": str(subscription_id),
```
```suggestion
                        "subscription_id": subscription_id,
```

This is also redundant with the message key itself (which is accessible to the consumer) but I suppose it's not a big deal for it to be in both places.

tkaemming commented on snuba/subscriptions/worker.py:
```diff
@@ -47,3 +50,24 @@ def flush_batch(self, batch: Sequence[Sequence[SubscriptionResult]]) -> None:
+                    "payload": {
+                        "subscription_id": str(subscription_id),
+                        "values": value.result["data"],
+                        "timestamp": value.task.timestamp.timestamp(),
```
We don't really deal with timestamps anywhere else that I can think of in the API. For consistency, I think it might be better for this to be `.isoformat()`.

tkaemming commented on snuba/subscriptions/worker.py:
```diff
@@ -47,3 +50,24 @@ def flush_batch(self, batch: Sequence[Sequence[SubscriptionResult]]) -> None:
+                    "version": 1,
+                    "payload": {
+                        "subscription_id": str(subscription_id),
+                        "values": value.result["data"],
```
I was thinking this would include the entire `ClickhouseQueryResult`.

wedamija replied:
Ahh, didn't quite follow your meaning. That seems fine, we might be able to use it for debugging on our end at some point too.

tkaemming commented on snuba/subscriptions/worker.py:
```diff
@@ -47,3 +50,24 @@ def flush_batch(self, batch: Sequence[Sequence[SubscriptionResult]]) -> None:
                 self.__producer.produce(self.__topic, result).result()
+
+
+class SubscriptionResultCodec(Codec[KafkaPayload, SubscriptionResult]):
```
It seems like this should go in `subscriptions.codecs`?

tkaemming replied:
Oops, disregard, didn't see other comment.

tkaemming reviewed (approved):
Some minor comments inline about the payload data for consistency with the HTTP interface, if it's going to be complicated to update the Sentry consumer to deal with those changes we can defer them for later but I think we should do them at some point.

tkaemming commented on snuba/subscriptions/worker.py:
```diff
@@ -47,3 +50,24 @@ def flush_batch(self, batch: Sequence[Sequence[SubscriptionResult]]) -> None:
                 self.__producer.produce(self.__topic, result).result()
+
+
+class SubscriptionResultCodec(Codec[KafkaPayload, SubscriptionResult]):
```
Yeah, this makes sense, we can follow up here after this last push to get everything wired up.

wedamija committed:
Update snuba/subscriptions/worker.py

Co-Authored-By: ted kaemming <ted@kaemming.com>

wedamija committed:
Update snuba/subscriptions/worker.py

Co-Authored-By: ted kaemming <ted@kaemming.com>

wedamija commented:
> Some minor comments inline about the payload data for consistency with the HTTP interface, if it's going to be complicated to update the Sentry consumer to deal with those changes we can defer them for later but I think we should do them at some point.

It's not too bad, I'm already updating the consumer to handle the updated format, so I'll make these changes there too.

wedamija committed:
Pass entire result
Pass timestamp as isoformat
Cleanups

wedamija committed:
Merge branch 'master' into danf/subscription_result_producer

tkaemming reviewed (approved):
This looks good, thanks for all the changes. Please go ahead and merge this after the tests pass.
